### PR TITLE
SONiC QOS Scheduler, WRED, Queue Yangs

### PR DESCRIPTION
**Pull request by AshokDaparthi**
Why I did it

Created SONiC Yang model for QOS
Tables: SCHEDULER, WRED_PROFILE, QUEUE table.
How I did it

Defined Yang models for QOS based on Guideline doc:
https://github.com/Azure/SONiC/blob/master/doc/mgmt/SONiC_YANG_Model_Guidelines.md
and
https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md
How to verify it

sonic_yang_models package build

Dependent pull requests: 
https://github.com/Azure/sonic-buildimage/pull/7752  - To modify platfrom files 
https://github.com/Azure/sonic-utilities/pull/1626   - DB migration 
https://github.com/Azure/sonic-swss/pull/1754    - swss change to remove ABNF format

### CODE REVIEW

**Comment by anshuv-mfst** [comment]
Hi @neethajohn - could you please take a look, thanks.

File: src/sonic-yang-models/yang-models/sonic-queue.yang
Comment (nazariig) [inline src/sonic-yang-models/yang-models/sonic-queue.yang:74]: @AshokDaparthi how do you expect to handle key reference notation from [swss-schema.md](https://github.com/Azure/sonic-swss/blob/master/doc/swss-schema.md)?

It seems that current YANG model won't be able to handle existing `config_db.json`:
**Example:**
```json
"QUEUE": {
    "Ethernet4|0": {
        "scheduler": "[SCHEDULER|scheduler.0]"
    },
    "Ethernet4|1": {
        "scheduler": "[SCHEDULER|scheduler.0]"
    },
    "Ethernet4|2": {
        "scheduler": "[SCHEDULER|scheduler.0]"
    },
    "Ethernet4|3": {
        "wred_profile": "[WRED_PROFILE|AZURE_LOSSLESS]",
        "scheduler": "[SCHEDULER|scheduler.1]"
    }
}

"SCHEDULER": {
    "scheduler.0": {
        "type": "DWRR",
        "weight": "14"
    },
    "scheduler.1": {
        "type": "DWRR",
        "weight": "15"
    }
}

"WRED_PROFILE": {
    "AZURE_LOSSLESS": {
        "red_max_threshold": "2097152",
        "wred_green_enable": "true",
        "ecn": "ecn_all",
        "green_min_threshold": "1048576",
        "red_min_threshold": "1048576",
        "wred_yellow_enable": "true",
        "yellow_min_threshold": "1048576",
        "green_max_threshold": "2097152",
        "green_drop_probability": "5",
        "yellow_max_threshold": "2097152",
        "wred_red_enable": "true",
        "yellow_drop_probability": "5",
        "red_drop_probability": "5"
    }
}

Which means that for scheduler/wred we expect to match:
1. scheduler -> scheduler.0
2. wred -> AZURE_LOSSLESS

and not:

1. scheduler -> SCHEDULER|scheduler.0
2. wred -> WRED_PROFILE|AZURE_LOSSLESS
```

Comment (nazariig) [reply src/sonic-yang-models/yang-models/sonic-queue.yang:74]: @praveen-li can you please elaborate what should be the appropriate YANG model for SONiC ABNF key reference?

How do we expect to handle this one?
```json
"BUFFER_PORT_EGRESS_PROFILE_LIST": {
    "Ethernet8": {
        "profile_list": "[BUFFER_PROFILE|egress_lossless_profile],[BUFFER_PROFILE|egress_lossy_profile]"
    },
```
```bash
root@sonic:/home/admin# redis-cli -n 4 HGETALL 'BUFFER_PORT_EGRESS_PROFILE_LIST|Ethernet8'
1) "profile_list"
2) "[BUFFER_PROFILE|egress_lossless_profile],[BUFFER_PROFILE|egress_lossy_profile]"
```

Comment (AshokDaparthi) [reply src/sonic-yang-models/yang-models/sonic-queue.yang:74]: @nazariig, @praveen-li  - We can discuss this over review call,  i am not sure pyang will take care leafref validation even if we give ABNF format. Else we should convert to leafref to pattern. But we will miss in this case leafref check's in YANG. As of now i am not sure is there anyway to resent in YANG.

File: src/sonic-yang-models/yang-models/sonic-scheduler.yang
Comment (rathnasabapathyv) [inline src/sonic-yang-models/yang-models/sonic-scheduler.yang:27]: Please update the description as Scheduler.

File: src/sonic-yang-models/yang-models/sonic-types.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-types.yang:118]: this is being discussed in another pr

File: src/sonic-yang-models/yang-models/sonic-wred-profile.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-wred-profile.yang:75]: is the operation check needed?

File: src/sonic-yang-models/yang-models/sonic-wred-profile.yang
Comment (rathnasabapathyv) [inline src/sonic-yang-models/yang-models/sonic-wred-profile.yang:61]: @AshokDaparthi : As discussed in the yang subgroup community call, please remove the default values from all these yang attributes.

**Comment by zhangyanzhao** [comment]
Depends on https://github.com/Azure/sonic-utilities/pull/1626

**Comment by zhangyanzhao** [comment]
Dependency Azure/sonic-utilities#1626 is merged

File: src/sonic-yang-models/yang-models/sonic-scheduler.yang
Comment (smaheshm) [inline src/sonic-yang-models/yang-models/sonic-scheduler.yang:90]: CIR is well know industry term. Mentioning the units is good enough in my opinion.

"CIR in bps for meter type 'bps' and pps for meter type 'pps'"

File: src/sonic-yang-models/yang-models/sonic-scheduler.yang
Comment (smaheshm) [inline src/sonic-yang-models/yang-models/sonic-scheduler.yang:107]: same comment as above.

File: src/sonic-yang-models/yang-models/sonic-queue.yang
Comment (smaheshm) [inline src/sonic-yang-models/yang-models/sonic-queue.yang:14]: port?

**Comment by smaheshm** [comment]
@AshokDaparthi Can you update the PR with suggestions. I see couple of open items, do they still need to get resolved? Are you waiting for inputs from the community?

**Comment by AshokDaparthi** [comment]
@smaheshm - Main issues blocking this PR is comments in sonic-queue.yang, Which needs DB format change.                      Below are PR raised   Azure/sonic-utilities#1626
Azure/sonic-swss#1754                                                                                                      
https://github.com/Azure/sonic-buildimage/pull/7752                                                                    
 https://github.com/Azure/sonic-mgmt/pull/3824  There is problems to pipeline testing because of all are interlinked changes and as per last call sonic-mgmt test cases needs both format in same code base based on release. It needs rework and find way to pass different config_db.json from yaml files with both new and old formats.

File: src/sonic-yang-models/yang-models/sonic-scheduler.yang
Comment (AshokDaparthi) [reply src/sonic-yang-models/yang-models/sonic-scheduler.yang:90]: @smaheshm - do you see any issue if we explain details in description?

Comment (smaheshm) [reply src/sonic-yang-models/yang-models/sonic-scheduler.yang:90]: No issues, leave it up to you.

File: src/sonic-yang-models/yang-models/sonic-queue.yang
Comment (AshokDaparthi) [reply src/sonic-yang-models/yang-models/sonic-queue.yang:74]: @nazariig - Back end changes to remove table ABNF format (example "[BUFFER_PROFILE|" )from field value is merged in latest master.

File: src/sonic-yang-models/yang-models/sonic-wred-profile.yang
Comment (AshokDaparthi) [reply src/sonic-yang-models/yang-models/sonic-wred-profile.yang]: removed operation based checks

File: src/sonic-yang-models/yang-models/sonic-wred-profile.yang
Comment (AshokDaparthi) [reply src/sonic-yang-models/yang-models/sonic-wred-profile.yang]: Done

**Comment by zhangyanzhao** [comment]
> @smaheshm - Main issues blocking this PR is comments in sonic-queue.yang, Which needs DB format change. Below are PR raised [Azure/sonic-utilities#1626](https://github.com/Azure/sonic-utilities/pull/1626) [Azure/sonic-swss#1754](https://github.com/Azure/sonic-swss/pull/1754) #7752 [Azure/sonic-mgmt#3824](https://github.com/Azure/sonic-mgmt/pull/3824) There is problems to pipeline testing because of all are interlinked changes and as per last call sonic-mgmt test cases needs both format in same code base based on release. It needs rework and find way to pass different config_db.json from yaml files with both new and old formats.

all the dependent PRs (1626/1754/7752/3824) are merged

**Comment by neethajohn** [comment]
> > @smaheshm - Main issues blocking this PR is comments in sonic-queue.yang, Which needs DB format change. Below are PR raised [Azure/sonic-utilities#1626](https://github.com/Azure/sonic-utilities/pull/1626) [Azure/sonic-swss#1754](https://github.com/Azure/sonic-swss/pull/1754) #7752 [Azure/sonic-mgmt#3824](https://github.com/Azure/sonic-mgmt/pull/3824) There is problems to pipeline testing because of all are interlinked changes and as per last call sonic-mgmt test cases needs both format in same code base based on release. It needs rework and find way to pass different config_db.json from yaml files with both new and old formats.
> 
> all the dependent PRs (1626/1754/7752/3824) are merged

We still have https://github.com/Azure/sonic-swss/pull/1937 pending and submodule update to ensure that the db reference changes are good. Once they are in, we can proceed with this PR

**Review by lguohan — APPROVED** [review]

**Comment by AshokDaparthi** [comment]
@smaheshm - Could you please help to see why it failing, i ran sonic-yang-models, and sonic-yang-mgmt python wheels. it success for me.

**Comment by smaheshm** [comment]
> @smaheshm - Could you please help to see why it failing, i ran sonic-yang-models, and sonic-yang-mgmt python wheels. it success for me.

looks like some process crashed, seen in other tests as well. I don't know if anyone's working on it. For I'd suggest update your branch and re-trigger the pipeline.

**Comment by zhangyanzhao** [comment]
Build failure need some help from people who can check the logs, looks like impacted by warmreboot test. Rajesh will re-push today

**Comment by AshokDaparthi** [comment]
@smaheshm, @lguohan - checks are passed finally, could you please review or merge it. since it already approved once.

**Review by smaheshm — APPROVED** [review]

**Pull request merged**
